### PR TITLE
sort: implement -k and -t support

### PR DESCRIPTION
**Pull request by miDeb**
For #1728.

This allows to specify keys after the -k flag and a custom field
separator using -t.

Not sure if I've missed something... For reference [GNU coreutils documentation about sort ](https://www.gnu.org/software/coreutils/manual/html_node/sort-invocation.html), [POSIX spec](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/sort.html).

### CODE REVIEW

**Comment by miDeb** [comment]
There's a test failure (of a test I added) saying `thread 'test_sort::test_keys_invalid_field_zero' panicked at 'Broken pipe (os error 32)', tests/common/util.rs:733:37`. Is this a problem with the test or something else?

**Comment by miDeb** [comment]
I think I want to implement the missing functionality first, before requesting a review

File: src/uu/sort/src/sort.rs [hunk 1]
```diff
@@ -87,13 +99,165 @@ impl Default for Settings {
...
+        let tokens = if !settings.selectors.is_empty() {
+            Some(tokenize(&input, settings.separator))
+        } else {
+            // If there are no selectors we will never need tokens, so don't calculate them in this case.
+            None
+        };
+        let transformed = transform(&input, settings);
+        let string;
+        let original;
+        if let Some(transformed) = transformed {
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:134]: please use tuples
https://doc.rust-lang.org/rust-by-example/primitives/tuples.html
to have
let (string, original) = if ...

File: src/uu/sort/src/sort.rs [hunk 2]
```diff
@@ -87,13 +99,165 @@ impl Default for Settings {
...
+// By default fields are separated by the first whitespace after non-whitespace.
+fn tokenize(line: &str, separator: Option<char>) -> Vec<Token> {
+    if let Some(separator) = separator {
+        tokenize_with_separator(line, separator)
+    } else {
+        tokenize_default(line)
+    }
+}
+
+fn tokenize_default(line: &str) -> Vec<Token> {
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:177]: can you please document this function?

File: src/uu/sort/src/sort.rs [hunk 3]
```diff
@@ -87,13 +99,165 @@ impl Default for Settings {
...
+            previous_was_whitespace = true;
+        } else {
+            previous_was_whitespace = false;
+        }
+    }
+    tokens.last_mut().unwrap().to = line.len();
+    tokens
+}
+
+fn tokenize_with_separator(line: &str, separator: char) -> Vec<Token> {
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:195]: can you please document this function?

File: src/uu/sort/src/sort.rs (hunk 1 above)
Comment (miDeb) [reply src/uu/sort/src/sort.rs:134]: Is this `destructuring_assingment`? I thought that wasn't stable yet (https://github.com/rust-lang/rust/issues/71126), but I'll try once I'm back to my computer

Comment (sylvestre) [reply src/uu/sort/src/sort.rs:134]: Oh, I might be wrong :)

Comment (miDeb) [reply src/uu/sort/src/sort.rs:134]: No, looks like you're right :)

Comment (sylvestre) [reply src/uu/sort/src/sort.rs:134]: Cool, this is rare :)

**Comment by electricboogie** [comment]
Just FYI, I've been working on sort too.  See most recently #2008.  Pleased to let you implement -k and -t, but be aware there is some functionality that I've added and I'd hate for us to needlessly conflict.  If you need my help with anything, please let me know.

**Comment by miDeb** [comment]
Thanks, I'll take a look at your PR, hopefully there aren't too many conflicts...
I think I figured out how to do it, but if you have any suggestions I'll be happy to take them.

**Comment by sylvestre** [comment]
```
---- test_sort::test_keys_invalid_field_zero stdout ----

current_directory_resolved: 

run: /home/travis/build/uutils/coreutils/target/debug/coreutils sort -k 0.1

thread 'test_sort::test_keys_invalid_field_zero' panicked at 'Broken pipe (os error 32)', tests/common/util.rs:742:37

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

Is failing twice in the CI

**Comment by miDeb** [comment]
I haven't been able to reproduce this locally, and I have no idea why this is failing for this test specifically...

**Comment by miDeb** [comment]
Well, that seems to have worked. There's one remaining failing test, saying
Error: Incompatible 'Cargo.lock' format; try `cargo +1.40.0 update`, which I guess isn't something I have caused because I haven't modified Cargo.lock at all.

**Comment by sylvestre** [comment]
Yeah, don't bother about minrustv :)

**Comment by sylvestre** [comment]
I will fix it here:
https://github.com/uutils/coreutils/pull/2013

**Comment by sylvestre** [comment]
Please let me know when it is ready :)

**Comment by miDeb** [comment]
This should be more or less ready. I don't know if I should wait for https://github.com/uutils/coreutils/pull/2008 to be merged first so I can rebase on top of it, as there are going to be some conflicts

**Comment by miDeb** [comment]
Thanks, I already expected that :)

**Comment by miDeb** [comment]
This should be ready now. Performance seems to have regressed slightly, but I don't think it's a big deal. Note that if you want to measure performance you should specify an outfile (`-o`), because I made writes to stdout buffered as a drive-by change, resulting in a clear perf win for redirecting stdout (`sort > outfile`).

**Comment by miDeb** [comment]
Spoke too soon, some tests are failing. Will fix them :)

**Comment by sylvestre** [comment]
could you please create and document how you did the benchmarking?
For example, create a 
`BENCHMARKING.md `
file
with where to get a data set (or generate it) and the commands to evaluate performances? (with hyperfine if possible)

thanks

**Comment by miDeb** [comment]
I added a benchmarking.md file. Below is a comparison of performance before and after my change:

wordlist: (`hyperfine "target/release/coreutils sort shuffled_wordlist.txt -o output.txt"`)
before:
`Time (mean ± σ):      26.1 ms ±   3.6 ms    [User: 64.3 ms, System: 9.6 ms]`
after:
`Time (mean ± σ):      27.2 ms ±   2.3 ms    [User: 107.5 ms, System: 12.7 ms]`

numbers: (`hyperfine "target/release/coreutils sort shuffled_numbers.txt -n -o output.txt"`)
before:
`Time (mean ± σ):      40.1 ms ±   3.8 ms    [User: 338.7 ms, System: 9.7 ms]`
after:
`Time (mean ± σ):      44.6 ms ±   3.6 ms    [User: 370.2 ms, System: 9.5 ms]`

**Pull request merged**